lib/CodeGen/MemoryAllocator.cpp

```diff
@@ -138,49 +154,626 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+  }
+  // Verify each ALLOC has an associated FREE following it.
+  // Verify each ALLOC has a unique handle.
+  std::list<Handle> allocHandleList;
+  for (auto allocIt = allocList.begin(); allocIt != allocList.end();
+       ++allocIt) {
+    if (!allocIt->alloc_) {
+      continue;
+    }
+    // Find a FREE instruction associated to this ALLOC.
+    auto allocHandle = allocIt->handle_;
+    bool freeFound = false;
+    for (auto it = allocIt; it != allocList.end(); ++it) {
+      if ((!it->alloc_) && (it->handle_ == allocHandle)) {
+        freeFound = true;
+        break;
+      }
+    }
+    if (!freeFound) {
+      return false;
+    }
+    // Verify ALLOC handle is unique.
+    auto handleIt =
+        std::find(allocHandleList.begin(), allocHandleList.end(), allocHandle);
+    if (handleIt != allocHandleList.end()) {
+      return false;
+    }
+    allocHandleList.emplace_back(allocHandle);
+  }
+  return true;
+}
+
+bool MemoryAllocator::verifySegments(
+    const std::list<Allocation> &allocList) const {
+  // Segments handles should match allocation handles.
+  // Segments sizes should match allocation sizes.
+  // Segments begin addresses must be aligned.
+  // Segments end addresses must not surpass the memory size.
+  for (const auto &alloc : allocList) {
+    if (!alloc.alloc_) {
+      continue;
+    }
+    auto it = handleToSegmentMap_.find(alloc.handle_);
+    if (it == handleToSegmentMap_.end()) {
+      return false;
+    }
+    Segment seg = it->second;
+    if (seg.size() != alloc.size_) {
+      return false;
+    }
+    if (seg.begin_ % alignment_) {
+      return false;
+    }
+    if (memorySize_ && seg.end_ > memorySize_) {
+      return false;
+    }
+  }
+  // Allocations which are simultaneously alive must be assigned non-overlapping
+  // segments.
+  std::list<Handle> liveHandleList;
+  for (const auto &alloc : allocList) {
+    auto allocHandle = alloc.handle_;
+    if (alloc.alloc_) {
+      // Verify current segment is not overlapping with the other live segments.
+      auto it = handleToSegmentMap_.find(alloc.handle_);
+      Segment seg = it->second;
+      for (const auto &liveHandle : liveHandleList) {
+        auto liveIt = handleToSegmentMap_.find(liveHandle);
+        Segment liveSeg = liveIt->second;
+        bool segOverlap = intervalsOverlap(seg.begin_, seg.end_, liveSeg.begin_,
+                                           liveSeg.end_);
+        if (segOverlap) {
+          return false;
+        }
+      }
+      // Add handle to live handles.
+      liveHandleList.emplace_back(allocHandle);
+    } else {
+      // Remove handle from live handles.
+      auto it =
+          std::find(liveHandleList.begin(), liveHandleList.end(), allocHandle);
+      assert(it != liveHandleList.end() && "Handle not found for removal!");
+      liveHandleList.erase(it);
+    }
+  }
+  assert(liveHandleList.empty() && "Inconsistent allocations!");
+  return true;
+}
+
+/// Buffer information structure.
```

Comment:
Maybe move the type definitions to the beginning of the file?

Reply:
I'm not sure I want to move it since this is only related to `allocateAll`. I want all the code related to `allocateAll` to be centralized for easier navigation.